targets/TARGET_Freescale/TARGET_MCUXpresso_MCUS/api/lp_ticker.c

```diff
@@ -71,102 +93,113 @@ void lp_ticker_init(void)
+        EnableIRQ(LPTMR0_IRQn);
+
+        lptmr_schedule = 0;
+
+        lp_ticker_inited = true;
+    } else {
+        /* In case of re-init we need to disable lp ticker interrupt. */
+        LPTMR_StopTimer(LPTMR0);
+
+        RTC_DisableInterrupts(RTC, kRTC_AlarmInterruptEnable);
+        RTC->TAR = 0; /* Write clears the IRQ flag */
+
+        lptmr_schedule = 0;
+    }
 }
 
 /** Read the current counter
  *
- * @return The current timer's counter value in microseconds
+ * @return The current timer's counter value in ticks
  */
 uint32_t lp_ticker_read(void)
 {
-    uint32_t sec, pre;
-
-    if (!lp_ticker_inited) {
-        lp_ticker_init();
-    }
-
-    sec = RTC->TSR; /* 32b: Seconds */
-    pre = RTC->TPR; /* 16b: Increments every 32.768kHz clock cycle (30us) */
-
-    /* Final value: 11b (4095) for sec and 21b for usec (pre can reach 1,000,000us which is close to 1<<20) */
-    uint32_t ret = (((sec & MAX_SEC_MASK) * SEC_IN_USEC) + (((uint64_t)pre * SEC_IN_USEC) / OSC32K_CLK_HZ));
-
-    return ret;
+    /* TPR is increments every 32.768 kHz clock cycle. The TSR increments when
+     * bit 14 of the TPR transitions from a logic one (32768 ticks - 1 sec).
+     * After that TPR starts counting from 0.
+     *
+     * count value is built as follows:
+     * count[0 - 14] - ticks (RTC->TPR)
+     * count[15 - 31] - seconds (RTC->TSR)
+     */
+    const uint32_t count = ((RTC->TSR << SEC_SHIFT) | (RTC->TPR & TICKS_MASK));
+
+    return count;
 }
 
 /** Set interrupt for specified timestamp
  *
- * @param timestamp The time in microseconds to be set
+ * @param timestamp The time in ticks to be set
  */
 void lp_ticker_set_interrupt(timestamp_t timestamp)
 {
-    uint32_t now_us, delta_us, delta_ticks;
-
-    if (!lp_ticker_inited) {
-        lp_ticker_init();
-    }
-
     lptmr_schedule = 0;
-    now_us = lp_ticker_read();
-    delta_us = timestamp >= now_us ? timestamp - now_us : (uint32_t)((uint64_t)timestamp + 0xFFFFFFFF - now_us);
 
-    /* Checking if LPTRM can handle this sleep */
-    delta_ticks = USEC_TO_COUNT(delta_us, CLOCK_GetFreq(kCLOCK_Er32kClk));
+    /* We get here absolute interrupt time-stamp in ticks which takes into account counter overflow.
+     * Since we use additional count-down timer to generate interrupt we need to calculate
+     * load value based on time-stamp.
+     */
+    const uint32_t now_ticks = lp_ticker_read();
+    uint32_t delta_ticks =
+            timestamp >= now_ticks ? timestamp - now_ticks : (uint32_t)((uint64_t) timestamp + 0xFFFFFFFF - now_ticks);
+
     if (delta_ticks == 0) {
-        /* The requested delay is less than the minimum resolution of this counter */
+        /* The requested delay is less than the minimum resolution of this counter. */
         delta_ticks = 1;
     }
 
     if (delta_ticks > MAX_LPTMR_SLEEP) {
-        /* Using RTC if wait time is over 16b (2s @32kHz) */
-        uint32_t delta_sec;
+        /* Using RTC if wait time is over 16b (2s @32kHz). */
```

Comment:
Does the RTC alarm interrupt need to be used here? It seems like this code could be made simpler and thus more robust by only using the LPTMR. You would just need to adjust the supported number of bits to 15 so delta_ticks will never be greater than MAX_LPTMR_SLEEP.

Reply:
I agree that this can be implemented using only RTC->TPR as 15 bit time counter and only LPTMR to generate interrupts, but on the other hand combining 32 bit time counter from RTC->TSR and RTC->TPR is also ok and we got wider counter range.
I suggest to leave 32-bit counter in this PR and consider change to 15-bit counter in another PR (I will create it) since we need to have some example implementations before meting with silicon partners and it looks like this version is stable.